pkg/controller/control.go

```diff
@@ -1287,3 +1329,79 @@ func (c *Controller) GetLatestMetics() *enginerpc.Metrics {
 func getAverageLatency(totalLatency, iops uint64) uint64 {
 	return totalLatency / iops
 }
+
+// tryFreeze attempts to bind mount an existing mount point to freezePoint, then freeze the file system from
+// freezePoint. tryFreeze returns booleans mounted and frozen so the caller can attempt to clean up depending on its
+// progress. tryFreeze does not return an error if there are no eligible mount points to freeze.
+func tryFreeze(devicePath string, freezePoint string, mounter mount.Interface, exec lhexec.ExecuteInterface,
+	log logrus.FieldLogger) (mounted, frozen bool, err error) {
+	// We do not need to switch to the host mount namespace to get mount points here. Usually, longhorn-engine runs in a
+	// container that has / bind mounted to /host with at least HostToContainer (rslave) propagation.
+	// - If it does not, we likely can't do a namespace swap anyway, since we don't have access to /host/proc.
+	// - If it does, we just need to know where in the container we can access the mount points to create a bind mount.
+	sourcePoints, err := mounter.List()
+	if err != nil {
+		log.WithError(err).Warn("Failed to list mount points while deciding whether or not to freeze")
+	}
+
+	for _, sourcePoint := range sourcePoints {
+		if sourcePoint.Device != devicePath {
+			continue // We are iterating through the list of all mounted file systems.
+		}
+		log.Debugf("Mounting file system to %v", freezePoint)
+		if err = attemptBindMountFileSystem(sourcePoint.Path, freezePoint, mounter); err != nil {
+			// There's no particular reason to think we will be successful with a different sourcePoint.
+			err = errors.Wrapf(err, "failed to bind mount file system from %v to %v", sourcePoint, freezePoint)
+			return
+		}
+		mounted = true
+
+		// We must verify it is still save to freeze the file system. If the source mount was unmounted by someone else
+		// before we bind mounted, the bind mount could refer to the root file system.
+		var device string
+		device, _, err = mount.GetDeviceNameFromMount(mounter, freezePoint)
+		if err != nil || device != devicePath {
+			// This would likely only happen if an upper layer unmounted the volume. If that is the case, we are
+			// probably being torn down. Trying other sourcePoints will likely lead to a pointless race.
+			err = errors.Wrapf(err, "cannot verify mount point %v refers to volume", freezePoint)
+			return
+		}
+
+		log.Infof("Freezing file system mounted at %v", freezePoint)
```

Comment:
I'm assuming we agree that this does not matter (since unmounting the source mount will not unmount the bind mount). I'm adding a comment to this effect.